File: cpp/839_相似字符串组_hard.cpp
class UnionFind {
private:
    vector<int> parent;
    vector<int> sz;
    int setCount;
public:
    UnionFind (int n): parent(n), sz(n, 1), setCount(n) {
        iota(parent.begin(), parent.end(), 0);
    }
    int find (int x) {
        return parent[x] == x ? x : parent[x] = find(parent[x]);
    }
    bool united (int x, int y) {
        x = find(x);
        y = find(y);
        if (x==y)
            return false;
        parent[y] = x;
        sz[x] += sz[y];
        setCount--;
        return true;
    }
    int get_setCount() {
        return setCount;
    }

};

class Solution {
public:
    int numSimilarGroups(vector<string>& strs) {
        int n = strs.size();
        auto uf = UnionFind(n);
        for(int i=0; i<n; i++)
            for(int j=i+1; j<n; j++) 
                if (equal(strs[i], strs[j]))
                    uf.united(i ,j);
        return uf.get_setCount();
    }
    bool equal(string a, string b) {
        int n = a.length();
        int count = 0;
        for(int i=0; i<n; i++) {
            if (a[i] != b[i])
                count++;
            if (count > 2)
                return false;
        }
        return true;
    }
};

File: python/839_相似字符串组_hard.py
class Solution:
    def numSimilarGroups(self, strs: List[str]) -> int:
        def equal(strx, stry):
            count = 0
            for i in range(len(strx)):
                if strx[i] != stry[i]:
                    count += 1
                if count > 2:
                    return False
            return True   
        n = len(strs)
        uf = UnionFind(n)
        for i in range(n):
            for j in range(i+1, n):
                if equal(strs[i], strs[j]):
                    uf.united(i, j)
        return uf.get_setCount()

class UnionFind():
    def __init__(self, n):
        self.n = n
        self.parent = [i for i in range(n)]
        self.sz = [1 for i in range(n)]
        self.setCount = n
    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]
    def united(self, x, y):
        x, y = self.find(x), self.find(y)
        if x == y:
            return False
        self.parent[y] = x
        self.sz[x] += self.sz[y]
        self.setCount -= 1
        return True
    def get_setCount(self):
        return self.setCount